### PR TITLE
Make observable inout copy/writeback/deinit orders more consistent

### LINKED ISSUE #15645: initialization and deinitialization order of formal temps
The current ordering of copy-initialization, write-back, and de-initialization for `in`, `inout`, and `out` formal temps does not seem right.

I would expect that `in` / `inout` / `out` arguments would:
 * have any required copy-initialization occur in argument order
 * have write-backs (for `inout`) in argument order
 * have deinitialization occur in reverse argument order

However this is not the case right now:
 * Copy-initialization for all `in` intent arguments is occuring before all `inout` arguments
 * `inout` argument copy initialziation is occuring in reverse argument order
 *  `in` and`inout` deinitialization is occuring in argument order (but `out` deinitialization occurs in reverse argument order)

Example test hidden below:

<details>

``` chapel
proc inout_inout(inout a: R, inout b: R) { }

proc test_inout_inout() {
  writeln();
  writeln("test_inout_inout");
  var a = new R(1);
  var b = new R(2);
  inout_inout(a, b);
  writeln(a, " ", b); // disabling copy elision for a and b
}
test_inout_inout();

proc in_inout(in a: R, inout b: R) { }

proc test_in_inout() {
  writeln();
  writeln("test_in_inout");
  var a = new R(1);
  var b = new R(2);
  in_inout(a, b);
  writeln(a, " ", b); // disabling copy elision for a and b
}
test_in_inout();

proc inout_in(inout a: R, in b: R) { }

proc test_inout_in() {
  writeln();
  writeln("test_inout_in");
  var a = new R(1);
  var b = new R(2);
  inout_in(a, b);
  writeln(a, " ", b); // disabling copy elision for a and b
}
test_inout_in();

proc in_in(in a: R, in b: R) { }

proc test_in_in() {
  writeln();
  writeln("test_in_in");
  var a = new R(1);
  var b = new R(2);
  in_in(a, b);
  writeln(a, " ", b); // disabling copy elision for a and b
}
test_in_in();

proc out_out(out a: R, out b: R) {
  a = new R(10);
  b = new R(20);
}

proc test_out_out() {
  writeln();
  writeln("test_out_out");
  var a = new R(1);
  var b = new R(2);
  out_out(a, b);
  writeln(a, " ", b); // disabling copy elision for a and b
}
test_out_out();

proc out_inout(out a: R, inout b: R) {
  a = new R(10);
  b = new R(20);
}

proc test_out_inout() {
  writeln();
  writeln("test_out_inout");
  var a = new R(1);
  var b = new R(2);
  out_inout(a, b);
  writeln(a, " ", b); // disabling copy elision for a and b
}
test_out_inout();

proc inout_out(inout a: R, out b: R) {
  a = new R(10);
  b = new R(20);
}

proc test_inout_out() {
  writeln();
  writeln("test_inout_out");
  var a = new R(1);
  var b = new R(2);
  inout_out(a, b);
  writeln(a, " ", b); // disabling copy elision for a and b
}
test_inout_out();








class C {
  var xx: int = 0;
}

record R {
  var x: int = 0;
  var ptr: shared C = new shared C(0);
  proc init() {
    this.x = 0;
    this.ptr = new shared C(0);
    writeln("init (default)");
  }
  proc init(arg:int) {
    this.x = arg;
    this.ptr = new shared C(arg);
    writeln("init ", arg, " ", arg);
  }
  proc init=(other: R) {
    this.x = other.x;
    this.ptr = new shared C(other.ptr.xx);
    writeln("init= ", other.x, " ", other.ptr.xx);
  }
  proc deinit() {
    writeln("deinit ", x, " ", ptr.xx);
  }
  proc toString() {
    return "(" + this.x:string + " " + this.ptr.xx:string + ")";
  }
  proc set1() {
    this.x = 1;
    this.ptr.xx = 1;
    return this;
  }
}
proc =(ref lhs:R, rhs:R) {
  writeln("lhs", lhs.toString(), " = rhs", rhs.toString());
  lhs.x = rhs.x;
  lhs.ptr = new shared C(rhs.ptr.xx);
}
```
</details>

### PR DESCRIPTION
**Pull request by mppf**
Resolves #15645.

This PR adjusts the compiler to implement support for `inout` by
splitting it into two arguments, an `in` argument and an `out` argument.
This allows the `in` part of `inout` to interleave with other `in`
arguments in a natural manner and allows the `out` part of `inout` to
interleave with other arguments in a natural manner.

At first, I tried to do this with one argument, but I ran into the
problems described here
https://github.com/chapel-lang/chapel/issues/15645#issuecomment-663640612
- namely that the write-back occurs in the body of the function and needs
a separate argument from the one receiving a copy - which is problematic
if the copy is occuring at the call site, which it needs to in order to
interleave properly with other `in` intent arguments.

The implementation approach here is to add a special formal, marked with
`FLAG_HIDDEN_FORMAL_INOUT`, just after each `inout` formal in normalize.
This special formal will handle the `out` part - and the original `inout`
formal will handle the `in` part. Then, adjust `ResolutionCandidate` to
not require that the hidden formal is passed. Then, adjust
`wrapAndCleanUpActuals` to handle the `in` part of an `inout` argument in
`handleInIntent`, and the `out` part in `handleDefaultArg` (which passes
the original argument to the `out` formal) and then in
`handleOutIntents`.  Since the `inout`/`in` part involves adding a temp
that obscures the original argument, the `fixHiddenInoutArg()` call sets
the actual for the hidden argument appropriately before running the `in`
part.
 
Besides adjusting `inout`, this PR adjusts `addLocalCopiesAndWritebacks`
to process formals in declaration order to make the order of
deinitialization more consistent. Previously, they were being processed
in the order they were stored in a map. While there, adjusted this
function to add elements to AST in regular order instead of reverse
order.

Lastly, I noticed that the promotion wrapper was calling special
functions to handle default arguments - addDefaultTokensAndReorder(),
replaceDefaultTokensWithDefaults(), fixDefaultArgumentsInWrapCall(). But
the promotion wrapper function will be resolved in its entirety and when
the call to the original function is resolved the default arguments will
be handled (along with `in` intent processing and lots of other stuff
that happens in wrapAndCleanUpActuals). Since default arguments will be
able to be handled at that later time, I removed the special handling of
default arguments in the promotion wrapper and the related support code.

Reviewed by @vasslitvinov - thanks!

- [x] primers pass with verify and valgrind
- [x] full local verify testing
- [x] full local futures testing

### CODE REVIEW

File: compiler/passes/checkParsed.cpp [hunk 1]
```diff
@@ -463,6 +463,16 @@ checkFunction(FnSymbol* fn) {
       numNonVoidReturns == 0) {
     USR_FATAL_CONT(fn, "function declared 'ref' but does not return anything");
   }
+
+  if (isIterator) {
+    for_formals(formal, fn) {
+      if (formal->intent == INTENT_OUT) {
+        USR_FATAL_CONT(formal, "out intent not yet supported for iterators");
```
Comment (vasslitvinov) [inline compiler/passes/checkParsed.cpp:470]: Going forward, I would like us to use English in cases like this, i.e. using "is" or another verb .
OK to leave as-is for this PR

File: compiler/resolution/functionResolution.cpp [hunk 2]
```diff
@@ -526,6 +526,34 @@ static bool fits_in_uint(int width, Immediate* imm) {
 }
```
Comment (vasslitvinov) [inline compiler/resolution/functionResolution.cpp:528]: Please add a brief comment explaining in particular what you mean by "in copy".

File: compiler/resolution/wrappers.cpp [hunk 3]
```diff
@@ -483,36 +437,6 @@ static void handleDefaultArg(FnSymbol *fn, CallExpr* call,
   body->flattenAndRemove();
 }
```
Comment (vasslitvinov) [inline compiler/resolution/wrappers.cpp:439]: Please add a note to the PR message about what allowed you to remove addDefaultTokensAndReorder(), replaceDefaultTokensWithDefaults(), fixDefaultArgumentsInWrapCall()

File: compiler/passes/checkParsed.cpp (hunk 1 above)
Comment (mppf) [reply compiler/passes/checkParsed.cpp:470]: I'll add *is*, no worries

**Pull request merged**